### ISSUE TITLE
Support Typecasting Function Parameters (Postgis ST_Transform Bug)

### ISSUE THREAD

**Issue by cprieger** [tgriesser/knex#2021]
I have come across a problem, with different solutions, but I would like to propose a feature that would allow the calling of functions to be typecasted in the parameterization. 

First the code: 
```    
function ST_Transform(shape, srid)
    {
        let sridParam = '?';

        if (_.isNumber(srid))
        {
            sridParam += '::integer';
        }

        return knex.raw(`st_transform(?, ${sridParam})`, [shape, srid]);
    },
```
The Problem: 
This is in a file called 'postgis.js' in my architecture. What it allows me to do is import postgis and call postgis.ST_Transform(geom, 4326). Originally this code was 1 line long that got wrapped up in a convenience function. That was until Postgis 2.3.0 dropped and they changed it so that ST_Transform now takes a string or an integer. This causes an error because parameterized raw statements do not get typecasted naturally.

The Solution: 
If knex had a property that was .executeFunction() that:
- Took a string representation of a function name. 
- An array of values to be passed in. 
- An array of typecastings to be used. 

It could be a very clean implementation for executing functions with correct types which has been requested multiple times from other issues. If we could start a discussion about a good way to accomplish this via code it would be appreciated and I would happily help with the implementation.

**Comment by elhigu** [comment]
I don't see much benefit in this compared to using `knex.raw` inside simple helper function. Could you give some more examples of complete queries, where this would make a big difference on usability?

**Comment by cprieger** [comment]
I don't think that you would see a big difference in usability. Because ultimately raw will do what you want it to. But I think it kind of falls in with the rest of library or so I thought? Personally I see it as a syntactic completion type thing (sorry if that doesn't play like I mean it to), since most databases offer some type of natural function or procedure support. 

Also thank you for forgiving my terrible op. I will try and clean up the formatting on it.

**Comment by elhigu** [comment]
@cprieger could you provide an example how the syntax would be when one uses this SQL function call wrapper? 

I have thought about it many times, but I've always ended up thinking that it is not worth of effort, because in this case knex.raw syntax would have been easier and more flexible than using some other wrapper method. 

So I'm interested to see how some real world query would look like the way you have thought this feature.

**Comment by cprieger** [comment]
Sure. Give me a bit to work on it some more. I want to make sure I include the weird dichotomy between PLPGSQL and T-SQL function execution. So i'll work up some good examples for you. Thanks for discussing this feature request.

**Comment by shai32** [comment]
I suggest doing something like
this is just pseudo code, it need works to make it work
```

knex.call = (...args) =>  {
 const functionName = args[0];
 const argumentsString = convert(_.last(args)); //remove first arguments and convert to sql query
 return knex.raw(`SELECT * FROM ${functionName}(argumentsString)`).then(res=> res.rows);
}

```
in this way we get a nice syntax to call [function](https://www.postgresql.org/docs/9.6/static/sql-syntax-calling-funcs.html)
if we a function like concat_lower_or_upper
we just do knex.call('concat_lower_or_upper', a ,b, true);
and we get the result in clean js

**Comment by elhigu** [comment]
@shai32 call function cannot be smart enough to invent correct conversions by itself. There is just not enough information for knex to be able to do that automatically. 

Knex doesn't know pretty much anything about what kind of schema or your DB have or what kind of stored procedures there are available.

**Comment by shai32** [comment]
@elhigu why knex need to know which stored procedures the DB has?
I am passing along the function name as a string. it just to convert it to select syntax.

also the args are pass along exactly in the order it was call from knex.call.

regarding the result conversation, its not a must feature if it can be done.

**Comment by elhigu** [comment]
@shai32 to be able to tell what type of parameters are valid for SQL function. That is the reason why  there is also typecastings given in the first proposal of the original post. If type casts are given explicitly then it is not necessary to know anything about the function that will be called.

But you mentioned:

> the call should be smart enough to convert to the right type.

and that is not possible without knowledge about the types that are acceptable for certain function.

**Comment by roytz** [comment]
@elhigu you can get the types of the parameters using javascript `type of` and convert them to sql syntax (for example, converting array to its sql syntax) and then call the function with the converted parameters. The function has already declared the parameters types it expects to receive so if there'll be any mismatch it would just throw an exception.

**Comment by elhigu** [comment]
@roytz Thanks I'am aware of javascript `typeof` operator. 

For example in postgres there are at least 3 different list/array syntaxes and it depends on context which one can be used. There are even many different string types (jsonb / normal) which are not always compatible without explicit casting. Javascript doesn't even have separate integer / float types and just plainly doesn't have enough type information to be able to do automatic mapping from jstype -> sqltype.

I already tried to explain that in my two earlier comments and I'm running out of words how do explain the problem further.

**Comment by roytz** [comment]
@elhigu didn't mean to upset you man, it's just that not until your last message did you mention that there might be multiple different postgres types which are represented using the same js type. Thank you.

**Comment by cprieger** [comment]
Didn't want to just disappear on this issue or have it go cold. I will try and get some work in this weekend on a good proposal for changes (maybe even a pull request Chronos willing). 

But in defense of @elhigu point wrapping this in your own Knex.raw helper class works just as well. 

This discussion is mainly for "Is there a clean implementation that could be adapted to meet Knex's standards and fit the overall model?" Which is actually fairly non-trivial when consideration is given to the different databases, typecastings, and syntactic sugars that can take place.

**Comment by jtlapp** [comment]
Were this implemented, it would be nice for the solution to play well with Typescript, so that Typescript users get static build-time type-checking.

**Comment by bonespiked** [comment]
The argument of "just use raw" is invalid at face value.  The whole point of knex is to provide a mid-level abstraction away from raw SQL whenever possible.  Adding in a common form if stored procedure calls are needed would be incredibly useful (particularly for PostGIS ).  It may not be fully supported for all platforms/use cases (see returning clause), but getting traction would drive the details for the edge cases.

**Comment by kibertoad** [comment]
@bonespiked Can't you use QueryBuilder.extend in your project if you need reusable version of this, given that you manage to get it work?

**Comment by bonespiked** [comment]
That still requires raw query development, just wrapping it inside knex as opposed to a utility function of my own.  I was just bumping this request up to have more 'native' support for stored procedures (since they aren't going anywhere)

**Comment by elhigu** [comment]
@bonespiked 

> The argument of "just use raw" is invalid at face value. The whole point of knex is to provide a mid-level abstraction away from raw SQL whenever possible. Adding in a common form if stored procedure calls are needed would be incredibly useful (particularly for PostGIS ). 

It is kind of funny that first you are talking about adding mid-level abstraction than may allow crossdb support and then you mention that it is particularly  useful for postgis that is not at all crossdb compatible (and this issue would not help with that). Mostly DB functions that you need to be calling are not crossdb compatible and if they are you can as easily use knex raw to call them like `knex.raw('min(??)', ['col_name'])` and you won't need to write separate code for each dialect. Having this function call interface will not help at all for that.

> That still requires raw query development, just wrapping it inside knex as opposed to a utility function of my own. 

That is default way for knex to handle DB specific functionality. If you have some specific crossdb feature to support in mind, please open feature request. Using raw in this case is perfectly valid and using raw with knex is not a hack it is normal part of query writing.

Though this feature in this issue might give at some point support for crossdb compatible type casting and semantic information for knex about SQL types that certain javascript values represent.

@ continuation to original issue:

As far as I can think of to be able to implement this we would need some kind of type wrapper classes to tell knex which database type (and if it is inbound or outbound parameter) each javascript parameter that is passed to the method are.

Something like this:

```
knex.funcCall('function_name', [
  knex.val(srId).integer(), 
  knex.ref('table.col'), 
  knex.val([1,2,3]).json(), 
  knex.out_var(result)
])

// should end up with something like
// knex.raw('function_name(?::integer, ??, ?::jsonb, <somehow do output> ?)', [srId, 'table.col', '[1,2,3]', result])
```

There has been actually an effort earlier to create this kind of SQL type wrapper methods, but it didn't get finished.

Also  that syntax above is just an example and I don't think it is particularly good api. Casting and type declaration wrappers could be separate to be able to allow casting also for knex.ref() objects.

And lastly that `out_var()` thing should be also investigated thoroughly to see how it works in different databases and see what kind of things should be supported there.

So basically this issue could be spit to 2 parts:
* sql typecasting and type  declaration helpers
* helper to create output variables for stored procedure

**Comment by mercmobily** [comment]
Thanks everyone for the discussion and examples.

Summary:

  - The underlying ask here is typed parameters for function calls (especially PostGIS). That’s valid and useful.
  - We’re not going to add a generic “function call API” to core that pretends  portability across dialects. The recommended pattern today is to use raw with explicit casts where needed, or wrap that in a small userland helper.

 Workaround today:

  - Postgres/PostGIS: use raw and explicit casts:
    knex.raw('ST_Transform(?::geometry, ?::integer)', [geom, srid])

Potential future direction:

  - If we add typed bindings, it should be a narrow, dialect‑aware feature (PG first), not a generic “call any function” abstraction. See #6287 (schema/ type descriptors) as a possible foundation for this.
  - If someone wants to champion a focused PG‑only typed‑binding helper (API + tests), please open an RFC/PR with a concrete design.

Given the lack of activity for nearly 6 years and no converged design, I’m closing this as out of scope for core. The guidance above is the recommended path; we can revisit with a concrete, minimal proposal tied to typed bindings.
